kubespawner/utils.py

```diff
@@ -27,3 +27,112 @@ def generate_hashed_slug(slug, limit=63, hash_length=6):
         prefix=slug[:limit - hash_length - 1],
         hash=slug_hash[:hash_length],
     ).lower()
+
+
+def update_k8s_model(target, source, logger=None, origin=None):
+    """
+    Takes a model instance such as V1PodSpec() and updates it with another
+    model representation. The origin parameter could be "extra_pod_config" for
+    example.
+    """
+    model = type(target)
+    if not hasattr(target, 'attribute_map'):
+        raise AttributeError("Attribute 'target' ({}) must be an object (such as 'V1PodSpec') with an attribute 'attribute_map'.".format(model.__name__))
+    if not isinstance(source, model) and not isinstance(source, dict):
+        raise AttributeError("Attribute 'source' ({}) must be an object of the same type as 'target' ({}) or a 'dict'.".format(type(source).__name__, model.__name__))
+
+    source_dict = _get_k8s_model_dict(model, source)
+    for key, value in source_dict.items():
+        if key not in target.attribute_map:
+            raise ValueError("The attribute 'source' ({}) contained '{}' not modeled by '{}'.".format(type(source).__name__, key, model.__name__))
+        if getattr(target, key):
+            if logger and origin:
+                logger.warning("Overriding KubeSpawner.{}'s value '{}' with '{}'.".format(origin, getattr(target, key), value))
+        setattr(target, key, value)
+
+    return target
+
+def get_k8s_model(model, model_dict):
+    """
+    Returns a model object from an model instance or represantative dictionary.
+    """
+    model_dict = copy.deepcopy(model_dict)
+
+    if isinstance(model_dict, model):
+        return model_dict
+    elif isinstance(model_dict, dict):
+        _map_dict_keys_to_model_attributes(model, model_dict)
+        return model(**model_dict)
+    else:
+        raise AttributeError("Expected object of type 'dict' (or '{}') but got '{}'.".format(model.__type__.__name__, model_dict.__type__.__name__))
+
+def _get_k8s_model_dict(model, obj):
+    """
+    Returns a model of dictionary kind
+    """
+    obj = copy.deepcopy(obj)
+
+    if isinstance(obj, model):
+        return obj.to_dict()
+    elif isinstance(obj, dict):
+        return _map_dict_keys_to_model_attributes(model, obj)
+    else:
+        raise AttributeError("Expected object of type '{}' (or 'dict') but got '{}'.".format(model.__type__.__name__, obj.__type__.__name__))
+
+def _map_dict_keys_to_model_attributes(model, model_dict):
+    """
+    Maps a dict's keys to the provided models attributes using its attribute_map
+    attribute. This is (always?) the same as converting camelCase to snake_case.
+    Note that the function will not influence nested object's keys.
+    """
+
+    for key in list(model_dict.keys()):
```

Comment:
Since i modify the model_dict while looping over its keys, I better loop over a copy of the keys rather than evaluating then while modifying them. So, I'm creating a new list from model_dict.keys() by passing it to list(), perhaps i should use copy() to make it more obvious?